packages/react-renderer-demo/src/pages/schema/validator-warnings.md

```diff
@@ -0,0 +1,75 @@
+import DocPage from '@docs/doc-page';
+
+<DocPage>
+
+# Validator warnings
+
+If you don't want to block the form from submitting when a validator fails, you can convert the validator into a warning validator.
+
+## 1. Enable warnings
+
+In the field, set `useWarnings` to `true`. This feature can impact performace, so it was made to be opt-in. Please, do not change this value run-time as it could introduce bugs because of React hooks ordering.
+
+```jsx
+{
+    component: 'text-field',
+    useWarnings: true,
+    ...
+}
+```
+
+## 2. Return warning
+
+### Object validator
+
+When you want to return an object validator as a function, just add `warning: true` to its configuration.
+
+```jsx
+{
+    ...,
+    validate: [{ type: 'required', warning: true }]
+}
+```
+
+### Function validator
+
+When using a function as the validator, the function has to return an object with the following format:
+
+```
+{
+    warning: true,
+    error: 'some message'
+}
+```
+
+## 3. Get warning
+
+A warning is stored in `meta.warning`. This value is specific for each field and fields with the same name do not share this. Also, this value is not reachable via `resolveProps` as it's computed just after the `resolveProps` function.
+
+```jsx
+const TextFieldWarning = (props) => {
+  const { input, meta, ...rest } = useFieldApi(props);
+  return (
+    <div>
+      <input {...input} {...rest} />
+      {meta.warning && <div id="warning">{meta.warning}</div>}
+    </div>
+  );
+};
+```
+
+### Convert to error
+
+By default, Data Driven Forms mappers are not supporting showing warnings. However you can convert the warning into an error - set `convertWarningToError` to `true`. This will replace `meta.error` with `meta.warning`, but `valid` etc. will be unchanged.
```

Comment:
I am not sure if this is a good approach. I think we can just dump the waring into a helper text. And have a priority 
```
error > warning
```
This would require some change to the compose validator and storing the warning to wait for the full result.

 I don't see a reason why would anybody start using soft validation and don't want to display it in the UI.

I think we should keep it as simple as possible. `convertWarningToError` is maybe just too much as well. As I understand it correctly, it only changes the "visual" representation of the field but the validation still fails. We have global props for that.

Reply:
`I think we can just dump the waring into a helper text.` I don't think that's a good idea. `helperText` has totally different meaning and you should be able to differentiate between helperText and warnings. helperText is also not-common props and you can have components without it.

`I don't see a reason why would anybody start using soft validation and don't want to display it in the UI.` It doesn't mean that users do not want to use soft validation without UI, it means that you can have different behavior for error / warning. This props was made to be able to use warnings in old components, that don't know what a warning is.

Reply:
> helperText has totally different meaning and you should be able to differentiate between helperText and warnings

This is actually not entirely true, if you look at current mappers they all use helper text to display errors. Pf4 actually even has a warning state and it consumes the helper text. We can add a simple CSS for MUI (obviously customizable) which will indicate the warning state. I think you are thinking of a description. But helper text should describe the current state of the input. So if there is a warning/error/success messages, that should be displayed inside helper text.

The only additional attribute for this feature should be the flag inside the validator or the result that's it.

> This props was made to be able to use warnings in old components, that don't know what a warning is.

Well yes, but it does not really work. The component will be reporting a warning state but actually, the form is in an error state. So it will give users false feedback. The fact that mappers don't know how to react to the state is a different issue and well, not all mappers will use this state. That is not an issue. The renderer should not try to migrate mappers for new features. We will have to adjust mappers so they can interpret the feature. I think we can simply start with MUI mapper. And PF4 because that actually has a warning state.

Reply:
Ok, I will remove this option at all. You are right, this should be handled by mappers separately, I just wanted to introduce a simple bridge until mapper will be ready.

We should really progress with the state manager, where we can implement warning right into the form state and all these issues will disappear. :snail: :boom:

Reply:
Yeah, I would love to get back to the state manager... Who knows maybe I'll have more time work on that now that we have the lockdown again :smile: